### PR TITLE
[FLINK-31311] Supports Bounded Watermark streaming read

### PR DESCRIPTION
**Pull request by JingsongLi**
There are some bound stream scenarios that require that stream reading can be ended. Generally speaking, the end event time is the better.

So in this ticket, supports writing the watermark to the snapshot and can specify the ending watermark when reading the stream.

### CODE REVIEW

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/table/source/snapshot/SnapshotEnumerator.java [hunk 1]
```diff
@@ -18,22 +18,78 @@
...
-     * the base files for the following incremental changes (or just return null if there are no
-     * base files).
+     * The first call to this method will produce a {@link DataFilePlan} containing the base files
+     * for the following incremental changes (or just return null if there are no base files).
      *
-     * <p>Following calls to this method will produce {@link DataTableScan.DataFilePlan}s containing
-     * incremental changed files. If there is currently no newer snapshots, null will be returned
-     * instead.
+     * <p>Following calls to this method will produce {@link DataFilePlan}s containing incremental
+     * changed files. If there is currently no newer snapshots, null will be returned instead.
```
Comment (yuzelin) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/table/source/snapshot/SnapshotEnumerator.java:36]: Need update comment.
1. produce Result ?
2. return null -> return FinishedResult

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/table/source/snapshot/SnapshotEnumerator.java [hunk 2]
```diff
@@ -18,22 +18,78 @@
...
+
+    /** Result for finished. */
+    class FinishedResult implements Result {
+
+        @Nullable
+        @Override
+        public DataFilePlan plan() {
+            throw new RuntimeException("This is a finished result.");
+        }
+    }
```
Comment (yuzelin) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/table/source/snapshot/SnapshotEnumerator.java:94]: I think maybe we can use `public static final FINISHED_RESULT = new Result{...}` instead.

**Review by yuzelin — APPROVED** [review]
LGTM.
I left some comments.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/table/source/snapshot/SnapshotEnumerator.java (hunk 2 above)
Comment (JingsongLi) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/table/source/snapshot/SnapshotEnumerator.java:94]: I still use `FinishedResult`, I think this is more clear to invokers.

**Pull request merged**